src/lib/operators/join_index.cpp

```diff
@@ -0,0 +1,358 @@
+          constexpr auto both_are_string_columns = left_is_string_column && right_is_string_column;
+
+          // clang-format off
+          if constexpr (neither_is_string_column || both_are_string_columns) {
+            auto iterable_left = create_iterable_from_column<LeftType>(typed_left_column);
+            auto iterable_right = create_iterable_from_column<RightType>(typed_right_column);
+
+            iterable_left.with_iterators([&](auto left_it, auto left_end) {
+              iterable_right.with_iterators([&](auto right_it, auto right_end) {
+                with_comparator(_scan_type, [&](auto comparator) {
+                  this->_join_two_columns(comparator, left_it, left_end, right_it, right_end, chunk_id_left,
+                                          chunk_id_right, left_matches);
+                });
+              });
+            });
+          }
+          // clang-format on
+        });
+      });
+    }
+
+    if (_is_outer_join) {
+      // add unmatched rows on the left for Left and Full Outer joins
+      for (ChunkOffset chunk_offset{0}; chunk_offset < left_matches.size(); ++chunk_offset) {
+        if (!left_matches[chunk_offset]) {
+          _pos_list_left->emplace_back(RowID{chunk_id_left, chunk_offset});
+          _pos_list_right->emplace_back(RowID{ChunkID{0}, INVALID_CHUNK_OFFSET});
+        }
+      }
+    }
+  }
+
+  // For Full Outer we need to add all unmatched rows for the right side.
+  // Unmatched rows on the left side are already added in the main loop above
+  if (_mode == JoinMode::Outer) {
+    for (ChunkID chunk_id_right = ChunkID{0}; chunk_id_right < right_table->chunk_count(); ++chunk_id_right) {
+      auto column_right = right_table->get_chunk(chunk_id_right).get_column(right_column_id);
+
+      resolve_data_and_column_type(right_data_type, *column_right, [&](auto right_type, auto& typed_right_column) {
+        using RightType = typename decltype(right_type)::type;
+
+        auto iterable_right = create_iterable_from_column<RightType>(typed_right_column);
+
+        iterable_right.for_each([&](const auto& right_value) {
+          const auto row_id = RowID{chunk_id_right, right_value.chunk_offset()};
+          if (!_right_matches.count(row_id)) {
+            _pos_list_left->emplace_back(RowID{ChunkID{0}, INVALID_CHUNK_OFFSET});
+            _pos_list_right->emplace_back(row_id);
+          }
+        });
+      });
+    }
+  }
+
+  // write output chunks
+  auto output_chunk = Chunk();
+
+  if (_mode == JoinMode::Right) {
+    _write_output_chunks(output_chunk, right_table, _pos_list_right);
+    _write_output_chunks(output_chunk, left_table, _pos_list_left);
+  } else {
+    _write_output_chunks(output_chunk, left_table, _pos_list_left);
+    _write_output_chunks(output_chunk, right_table, _pos_list_right);
+  }
+
+  _output_table->emplace_chunk(std::move(output_chunk));
+}
+
+void JoinIndex::_write_output_chunks(Chunk& output_chunk, const std::shared_ptr<const Table> input_table,
```

Comment:
Could you please check if we have this function in every join, implementation? If so, please extract and reuse.

Reply:
Function is probably better named `_write_output_chunk` with the trailing `s`.

Reply:
The function is only present in the `JoinNestedLoop`, not in `JoinSortMerge` or `JoinHash`.
We might think about extracting common methods for the NLJ and Index-join, as we currently also reuse the `_create_table_structure` function.